// @plasmicapp/data-sources/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/index.tsx
var index_exports = {};
__export(index_exports, {
  Fetcher: () => Fetcher,
  FetcherMeta: () => FetcherMeta,
  _StatefulQueryResult: () => StatefulQueryResult,
  deriveFieldConfigs: () => deriveFieldConfigs,
  executePlasmicDataOp: () => executePlasmicDataOp,
  executeServerQuery: () => executeServerQuery,
  makeCacheKey: () => makeCacheKey,
  makeQueryCacheKey: () => makeQueryCacheKey,
  normalizeData: () => normalizeData,
  throwIfPlasmicUndefinedDataError: () => throwIfPlasmicUndefinedDataError,
  unstable_executePlasmicQueries: () => executePlasmicQueries,
  unstable_usePlasmicQueries: () => usePlasmicQueries,
  unstable_wrapDollarQueriesForMetadata: () => wrapDollarQueriesForMetadata,
  useNormalizedData: () => useNormalizedData,
  usePlasmicDataConfig: () => import_query5.usePlasmicDataConfig,
  usePlasmicDataMutationOp: () => usePlasmicDataMutationOp,
  usePlasmicDataOp: () => usePlasmicDataOp,
  usePlasmicInvalidate: () => usePlasmicInvalidate
});
module.exports = __toCommonJS(index_exports);

// src/common.ts
var import_query = require("@plasmicapp/query");
var ph = __toESM(require("@plasmicapp/host"));
var import_react = __toESM(require("react"));
function isPlasmicUndefinedDataErrorPromise(x) {
  return !!x && typeof x === "object" && (x == null ? void 0 : x.plasmicType) === "PlasmicUndefinedDataError";
}
function throwIfPlasmicUndefinedDataError(err) {
  if (isPlasmicUndefinedDataErrorPromise(err)) {
    throw err;
  }
}
function tagPlasmicUndefinedDataErrorPromise(promise) {
  promise.plasmicType = "PlasmicUndefinedDataError";
  promise.message = "Query is not done";
}
function untagPlasmicUndefinedDataErrorPromise(promise) {
  delete promise.plasmicType;
  delete promise.message;
}
function mkUndefinedDataProxy(promiseRef, fetchAndUpdateCache) {
  let fetchAndUpdatePromise = void 0;
  return new Proxy(
    {},
    {
      get: (_target, prop) => {
        if (prop === "isPlasmicUndefinedDataProxy") {
          return true;
        }
        if (!fetchAndUpdateCache) {
          return void 0;
        }
        const doFetchAndUpdate = () => {
          if (!fetchAndUpdatePromise) {
            fetchAndUpdatePromise = fetchAndUpdateCache().finally(() => {
              fetchAndUpdatePromise = void 0;
            });
          }
          return fetchAndUpdatePromise;
        };
        const promise = (
          // existing fetch
          promiseRef.fetchingPromise || // No existing fetch, so kick off a fetch
          doFetchAndUpdate()
        );
        promise.plasmicType = "PlasmicUndefinedDataError";
        promise.message = `Cannot read property ${String(
          prop
        )} - data is still loading`;
        throw promise;
      }
    }
  );
}
var isRSC = import_react.default.isRSC;
var reactMajorVersion = +import_react.default.version.split(".")[0];
var enableLoadingBoundaryKey = "plasmicInternalEnableLoadingBoundary";
var PRE_FETCHES = /* @__PURE__ */ new Map();
function usePlasmicFetch(key, resolvedParams, fetcherFn, resultMapper, undefinedDataProxyFields, opts) {
  var _a, _b;
  const enableLoadingBoundary = !!((_b = (_a = ph.useDataEnv) == null ? void 0 : _a()) == null ? void 0 : _b[enableLoadingBoundaryKey]);
  const { mutate, cache } = isRSC ? {} : (0, import_query.usePlasmicDataConfig)();
  const isNullParams = !resolvedParams;
  const isWaitingOnDependentQuery = isPlasmicUndefinedDataErrorPromise(resolvedParams);
  const fetchingData = import_react.default.useMemo(
    () => ({
      fetchingPromise: void 0
    }),
    [key]
  );
  const fetcher = import_react.default.useMemo(
    () => () => {
      if (!key) {
        throw new Error(`Fetcher should never be called without a proper key`);
      }
      if (fetchingData.fetchingPromise) {
        return fetchingData.fetchingPromise;
      }
      if (key && PRE_FETCHES.has(key)) {
        const existing = PRE_FETCHES.get(key);
        fetchingData.fetchingPromise = existing;
        return existing;
      }
      const fetcherPromise = fetcherFn(resolvedParams);
      fetchingData.fetchingPromise = fetcherPromise;
      if (key) {
        PRE_FETCHES.set(key, fetcherPromise);
        fetcherPromise.then(
          () => {
            PRE_FETCHES.delete(key);
          },
          () => {
            PRE_FETCHES.delete(key);
          }
        );
      }
      return fetcherPromise;
    },
    [key, fetchingData]
  );
  const dependentKeyDataErrorPromise = isPlasmicUndefinedDataErrorPromise(
    resolvedParams
  ) ? resolvedParams : void 0;
  const fetchAndUpdateCache = import_react.default.useMemo(() => {
    if (!key && !dependentKeyDataErrorPromise) {
      return void 0;
    }
    return () => {
      if (fetchingData.fetchingPromise) {
        return fetchingData.fetchingPromise;
      }
      if (dependentKeyDataErrorPromise) {
        return dependentKeyDataErrorPromise;
      }
      if (!key) {
        throw new Error(`Expected key to be non-null`);
      }
      const cached = cache == null ? void 0 : cache.get(key);
      if (cached) {
        return Promise.resolve(cached);
      }
      const cachedError = cache == null ? void 0 : cache.get(`$swr$${key}`);
      if (cachedError) {
        return Promise.reject(cachedError.error);
      }
      const fetcherPromise = new Promise((resolve, reject) => {
        setTimeout(() => {
          fetcher().then(resolve, reject);
        }, 1);
      });
      if (!isRSC) {
        fetcherPromise.then((data2) => {
          mutate == null ? void 0 : mutate(key, data2);
        }).catch((err) => {
          var _a2;
          const keyInfo = key ? "$swr$" + key : "";
          cache == null ? void 0 : cache.set(keyInfo, __spreadProps(__spreadValues({}, (_a2 = cache == null ? void 0 : cache.get(keyInfo)) != null ? _a2 : {}), { error: err }));
        });
      }
      return fetcherPromise;
    };
  }, [fetcher, fetchingData, cache, key, dependentKeyDataErrorPromise]);
  const res = (0, import_query.useMutablePlasmicQueryData)(key, fetcher, {
    fallbackData: opts == null ? void 0 : opts.fallbackData,
    shouldRetryOnError: false,
    // If revalidateIfStale is true, then if there's a cache entry with a key,
    // but no mounted hook with that key yet, and when the hook mounts with the key,
    // swr will revalidate. This may be reasonable behavior, but for us, this
    // happens all the time -- we prepopulate the cache with proxy-invoked fetch,
    // sometimes before swr had a chance to run the effect.  So we turn off
    // revalidateIfStale here, and just let the user manage invalidation.
    revalidateIfStale: false
  });
  const { data, error, isLoading } = res;
  if (fetchingData.fetchingPromise != null && data !== void 0) {
    fetchingData.fetchingPromise = void 0;
  }
  return import_react.default.useMemo(() => {
    const result = resultMapper(res);
    if (!(opts == null ? void 0 : opts.noUndefinedDataProxy) && reactMajorVersion >= 18 && enableLoadingBoundary && (isLoading || isNullParams || isWaitingOnDependentQuery) && undefinedDataProxyFields.every((field) => result[field] === void 0)) {
      undefinedDataProxyFields.forEach((field) => {
        if (field === "error") {
          return;
        }
        result[field] = mkUndefinedDataProxy(fetchingData, fetchAndUpdateCache);
      });
    }
    return result;
  }, [
    isNullParams,
    isWaitingOnDependentQuery,
    data,
    error,
    isLoading,
    opts == null ? void 0 : opts.noUndefinedDataProxy,
    enableLoadingBoundary,
    fetchingData,
    fetchAndUpdateCache
  ]);
}

// src/serverQueries/client.ts
var import_query2 = require("@plasmicapp/query");
var React3 = __toESM(require("react"));

// src/utils.ts
function noopFn() {
}
function pick(obj, ...keys) {
  const res = {};
  for (const key of keys) {
    if (key in obj) {
      res[key] = obj[key];
    }
  }
  return res;
}
var tuple = (...args) => args;
function mkIdMap(xs) {
  return new Map(xs.map((x) => tuple(x.id, x)));
}
function notNil(x) {
  return x !== null && x !== void 0;
}
function withoutNils(xs) {
  return xs.filter(notNil);
}
function mapRecordEntries(callback, record1, record2, record3) {
  return Object.entries(record1).map(([k, v1]) => {
    const v2 = record2 == null ? void 0 : record2[k];
    const v3 = record3 == null ? void 0 : record3[k];
    return callback(
      k,
      v1,
      v2,
      v3
    );
  });
}
function mapRecords(callback, record1, record2, record3) {
  return Object.fromEntries(
    mapRecordEntries(
      (k, v1, v2, v3) => [k, callback(k, v1, v2, v3)],
      record1,
      record2,
      record3
    )
  );
}

// src/serverQueries/common.ts
var import_react2 = __toESM(require("react"));

// src/serverQueries/makeQueryCacheKey.ts
function makeQueryCacheKey(id, params) {
  return `${id}:${safeStableStringify(params)}`;
}
var shortPlasmicPrefix = "\u03C1";
function safeStableStringify(unstableValue) {
  const stableValue = sortObjectsDeep(unstableValue, /* @__PURE__ */ new Map());
  const visitedPaths = /* @__PURE__ */ new Map();
  return JSON.stringify(stableValue, function(key, value) {
    switch (typeof value) {
      case "undefined":
        return `${shortPlasmicPrefix}:UNDEFINED`;
      case "function":
        return `${shortPlasmicPrefix}:FUNCTION:${value.name}`;
      case "symbol":
        return `${shortPlasmicPrefix}:SYMBOL:${value.description}`;
      case "bigint":
        return value.toString();
      case "object":
        if (value !== null) {
          const cyclePath = visitedPaths.get(value);
          if (cyclePath) {
            return `${shortPlasmicPrefix}:REF:${cyclePath}`;
          }
          const parentPath = visitedPaths.get(this);
          const valuePath = parentPath === void 0 ? "$" : `${parentPath}.${key}`;
          visitedPaths.set(value, valuePath);
        }
        return value;
      default:
        return value;
    }
  });
}
function sortObjectsDeep(value, visitedObjects) {
  if (typeof value !== "object" || value === null) {
    return value;
  }
  const visitedValue = visitedObjects.get(value);
  if (visitedValue) {
    return visitedValue;
  }
  if (typeof value.toJSON === "function") {
    return sortObjectsDeep(value.toJSON(), visitedObjects);
  }
  if (Array.isArray(value)) {
    const newArr = [];
    visitedObjects.set(value, newArr);
    value.forEach((item, key) => {
      newArr[key] = sortObjectsDeep(item, visitedObjects);
    });
    return newArr;
  } else {
    const newObj = {};
    visitedObjects.set(value, newObj);
    Object.keys(value).sort().forEach((key) => {
      newObj[key] = sortObjectsDeep(value[key], visitedObjects);
    });
    return newObj;
  }
}

// src/serverQueries/common.ts
function createDollarQueries(queryNames) {
  return Object.fromEntries(
    queryNames.map((queryName) => {
      return [queryName, new StatefulQueryResult()];
    })
  );
}
var StatefulQueryResult = class {
  constructor() {
    __publicField(this, "current");
    __publicField(this, "settable");
    __publicField(this, "listeners", /* @__PURE__ */ new Set());
    this.current = { state: "initial", key: null };
    this.settable = new SettablePromise();
    tagPlasmicUndefinedDataErrorPromise(this.settable.promise);
  }
  transitionState(next) {
    const prev = this.current;
    if (prev.state === next.state && prev.key === next.key && ("data" in prev ? prev.data : void 0) === ("data" in next ? next.data : void 0) && ("error" in prev ? prev.error : void 0) === ("error" in next ? next.error : void 0)) {
      return;
    }
    if (prev.key !== null && // the promise was already bound to a key AND
    // NOT the loading -> done transition with the same key
    !(prev.key === next.key && prev.state === "loading" && next.state === "done")) {
      this.settable = new SettablePromise();
      tagPlasmicUndefinedDataErrorPromise(this.settable.promise);
    }
    this.current = next;
    this.notifyListeners(prev);
  }
  addListener(listener) {
    this.listeners.add(listener);
  }
  removeListener(listener) {
    this.listeners.delete(listener);
  }
  notifyListeners(prev) {
    for (const listener of this.listeners) {
      listener(this.current, prev);
    }
  }
  reset() {
    this.transitionState({
      state: "initial",
      key: null
    });
  }
  loadingPromise(key, promise) {
    this.transitionState({
      state: "loading",
      key
    });
    promise.then(
      (value) => this.resolvePromise(key, value),
      (err) => this.rejectPromise(key, err)
    );
  }
  resolvePromise(key, data) {
    if (this.current.key === null || this.current.key === key) {
      this.transitionState({
        state: "done",
        key,
        data
      });
      this.settable.resolve(data);
      untagPlasmicUndefinedDataErrorPromise(this.settable.promise);
    }
  }
  /**
   * Reject is allowed if:
   * 1) no key / state is initial, which means params resolution failed
   * 2) key / state is loading, which means we need to check the keys match
   */
  rejectPromise(key, error) {
    if (this.current.key === null || this.current.key === key) {
      this.transitionState({
        state: "done",
        key,
        error
      });
      this.settable.promise.catch(noopFn);
      this.settable.reject(error);
      untagPlasmicUndefinedDataErrorPromise(this.settable.promise);
    }
  }
  toJSON() {
    return this.current;
  }
  get key() {
    return this.current.key;
  }
  get data() {
    if (this.current.state === "done") {
      if ("data" in this.current) {
        return this.current.data;
      } else {
        throw this.current.error;
      }
    } else {
      throw this.settable.promise;
    }
  }
  get isLoading() {
    return this.current.state === "initial" || this.current.state === "loading";
  }
  getDoneResult() {
    return __async(this, null, function* () {
      yield this.settable.promise;
      return this;
    });
  }
};
function safeExec(tryData, ifUndefined, ifError) {
  const result = safeExecResult(tryData);
  if ("promise" in result) {
    return ifUndefined(result.promise);
  } else if ("error" in result) {
    return ifError(result.error);
  } else {
    return result.data;
  }
}
function safeExecResult(tryData) {
  try {
    return { data: tryData() };
  } catch (err) {
    if (isPlasmicUndefinedDataErrorPromise(err)) {
      return { promise: err };
    } else {
      return { error: err };
    }
  }
}
function assertUnexpectedNodeType(x) {
  throw new Error(`Unexpected node type: ${x}`);
}
function resolveParams(queryId, params) {
  return safeExec(
    () => {
      const resolvedParams = params();
      const cacheKey = makeQueryCacheKey(queryId, resolvedParams);
      return {
        status: "ready",
        resolvedParams,
        cacheKey
      };
    },
    (promise) => ({
      status: "blocked",
      promise
    }),
    (err) => ({
      status: "error",
      error: new Error("Error resolving function params", { cause: err })
    })
  );
}
function wrapDollarQueriesForMetadata($queries, ifUndefined, ifError) {
  return wrapDollarQueriesWithFallbacks(
    $queries,
    ifUndefined != null ? ifUndefined : (() => "\u2026"),
    ifError != null ? ifError : (() => "[ERROR]")
  );
}
function wrapDollarQueriesWithFallbacks($queries, ifUndefined, ifError) {
  return mapRecords(
    (_queryName, $query) => new FallbackQueryResult($query, ifUndefined, ifError),
    $queries
  );
}
var FallbackQueryResult = class {
  constructor($query, ifUndefined, ifError) {
    __publicField(this, "$query", $query);
    __publicField(this, "ifUndefined", ifUndefined);
    __publicField(this, "ifError", ifError);
  }
  get key() {
    return this.$query.key;
  }
  get data() {
    return safeExec(
      () => this.$query.data,
      (promise) => createConstantProxy(this.ifUndefined(promise)),
      (err) => createConstantProxy(this.ifError(err))
    );
  }
  get isLoading() {
    return this.$query.isLoading;
  }
};
function createConstantProxy(constant) {
  const constantProxy = new Proxy(
    {},
    {
      get(_target, prop) {
        return prop === Symbol.toPrimitive ? () => constant : constantProxy;
      }
    }
  );
  return constantProxy;
}
var SyncPromise = class {
  constructor(promise) {
    __publicField(this, "promise", promise);
    __publicField(this, "result");
    promise.then(
      (value) => {
        this.result = {
          state: "resolved",
          value
        };
      },
      (error) => {
        this.result = {
          state: "rejected",
          error
        };
      }
    );
  }
};
var SettablePromise = class {
  constructor() {
    __publicField(this, "promise");
    __publicField(this, "_resolve");
    __publicField(this, "_reject");
    this.promise = new Promise((resolve, reject) => {
      this._resolve = resolve;
      this._reject = reject;
    });
  }
  resolve(value) {
    this._resolve(value);
  }
  reject(error) {
    this._reject(error);
  }
};
function usePrevious(value) {
  const ref = import_react2.default.useRef(void 0);
  const prev = ref.current;
  ref.current = value;
  return prev;
}
function createInitial$State($ctx, $props, $q, stateSpecs) {
  const root = {};
  for (const stateSpec of stateSpecs) {
    if (stateSpec.path.includes("[]")) {
      continue;
    }
    const parts = stateSpec.path.split(".");
    const parentPath = parts.slice(0, parts.length - 1);
    const leaf = parts[parts.length - 1];
    let parent = root;
    for (const part of parentPath) {
      if (!(part in parent)) {
        parent[part] = {};
      }
      parent = parent[part];
    }
    if (stateSpec.valueProp) {
      parent[leaf] = $props[stateSpec.valueProp];
    } else if ("initVal" in stateSpec) {
      parent[leaf] = stateSpec.initVal;
    } else if (stateSpec.initFunc) {
      const initFunc = stateSpec.initFunc;
      let cached;
      Object.defineProperty(parent, leaf, {
        get: () => {
          if (cached) {
            return cached.value;
          }
          const value = initFunc({
            $ctx,
            $props,
            $q,
            $state: root,
            $refs: {},
            $queries: {}
          });
          cached = { value };
          return value;
        },
        enumerable: true,
        configurable: true
      });
    }
  }
  return root;
}

// src/serverQueries/client.ts
var GLOBAL_CACHE = /* @__PURE__ */ new Map();
function usePlasmicQueries(tree, $ctx, $props, $state) {
  var _a;
  const wrappedQueries = React3.useMemo(() => wrapQueries(tree.queries), [tree]);
  const $queries = React3.useMemo(
    () => createDollarQueries(Object.keys(tree.queries)),
    [tree]
  );
  const $queryStates = $queries;
  if (!$state) {
    $state = createInitial$State($ctx, $props, $queryStates, tree.stateSpecs);
  }
  const { fallback: prefetchedCache, cache: swrCache } = (0, import_query2.usePlasmicDataConfig)();
  const paramsResults = {};
  const executionCtx = {
    $ctx,
    $props,
    $q: $queryStates,
    $state
  };
  const prevWrappedQueries = usePrevious(wrappedQueries);
  let consistent = prevWrappedQueries === void 0 || Object.is(prevWrappedQueries, wrappedQueries);
  mapRecords(
    (queryName, $query, query) => {
      if (!consistent || $query.current.state === "initial") {
        return;
      }
      const paramsResult = resolveParams(
        query.id,
        () => query.args(executionCtx)
      );
      paramsResults[queryName] = paramsResult;
      if (paramsResult.status === "blocked") {
        consistent = false;
      } else if (paramsResult.status === "error" && $query.current.key !== null) {
        consistent = false;
      } else if (paramsResult.status === "ready" && paramsResult.cacheKey !== $query.current.key) {
        consistent = false;
      }
    },
    $queryStates,
    wrappedQueries
  );
  if (!consistent) {
    mapRecords((_queryName, $query) => {
      $query.reset();
    }, $queryStates);
    for (const k of Object.keys(paramsResults)) {
      delete paramsResults[k];
    }
  }
  const stopRef = React3.useRef();
  (_a = stopRef.current) == null ? void 0 : _a.call(stopRef);
  let stopped = false;
  const stop = new Promise((resolve) => {
    stopRef.current = () => {
      stopped = true;
      resolve();
    };
  });
  React3.useEffect(() => () => {
    var _a2;
    return (_a2 = stopRef.current) == null ? void 0 : _a2.call(stopRef);
  }, []);
  const loop = () => __async(null, null, function* () {
    while (true) {
      initPlasmicQueriesSync(
        $queryStates,
        wrappedQueries,
        paramsResults,
        executionCtx,
        prefetchedCache,
        swrCache
      );
      const loadingQueries = mapRecordEntries((_queryName, $query) => {
        if ($query.isLoading) {
          return $query.getDoneResult();
        } else {
          return null;
        }
      }, $queryStates).filter(notNil);
      if (loadingQueries.length === 0) {
        break;
      }
      yield Promise.race([stop, ...loadingQueries]);
      if (stopped) {
        break;
      }
    }
  });
  loop().catch(noopFn);
  mapRecords(
    (queryName, $query, query) => {
      usePlasmicQuery($query, query, paramsResults[queryName]);
    },
    $queryStates,
    wrappedQueries
  );
  return $queries;
}
function wrapQueries(queries) {
  return mapRecords((_queryName, query) => {
    const wrappedFn = (...args) => {
      const cacheKey = makeQueryCacheKey(query.id, args);
      const cached = GLOBAL_CACHE.get(cacheKey);
      if (cached) {
        return cached.promise;
      }
      const promise = (0, import_query2.wrapLoadingFetcher)(query.fn)(...args);
      GLOBAL_CACHE.set(cacheKey, new SyncPromise(promise));
      return promise;
    };
    return {
      id: query.id,
      fn: wrappedFn,
      args: query.args
    };
  }, queries);
}
function initPlasmicQueriesSync($queries, queries, paramsResults, executionCtx, prefetchedCache, clientCache) {
  let anySettled;
  do {
    anySettled = false;
    mapRecords(
      (queryName, $query, query) => {
        if ($query.current.state !== "initial") {
          return;
        }
        const paramsResult = resolveParams(
          query.id,
          () => query.args(executionCtx)
        );
        paramsResults[queryName] = paramsResult;
        if (paramsResult.status === "error") {
          $query.rejectPromise(null, paramsResult.error);
          anySettled = true;
          return;
        } else if (paramsResult.status === "blocked") {
          return;
        }
        const cacheKey = makeQueryCacheKey(
          query.id,
          paramsResult.resolvedParams
        );
        if (cacheKey in prefetchedCache) {
          $query.resolvePromise(cacheKey, prefetchedCache[cacheKey]);
          anySettled = true;
          return;
        }
        const clientCacheData = clientCache.get(cacheKey);
        if (clientCacheData !== void 0) {
          $query.resolvePromise(cacheKey, clientCacheData);
          anySettled = true;
          return;
        }
        const clientCachedPromise = GLOBAL_CACHE.get(cacheKey);
        if (clientCachedPromise == null ? void 0 : clientCachedPromise.result) {
          if (clientCachedPromise.result.state === "resolved") {
            $query.resolvePromise(cacheKey, clientCachedPromise.result.value);
          } else {
            $query.rejectPromise(cacheKey, clientCachedPromise.result.error);
          }
          anySettled = true;
          return;
        }
        $query.loadingPromise(
          cacheKey,
          query.fn(...paramsResult.resolvedParams)
        );
      },
      $queries,
      queries
    );
  } while (anySettled);
}
function usePlasmicQuery($query, query, paramsResult) {
  const $queryState = $query;
  const { key, fetcher } = React3.useMemo(() => {
    switch (paramsResult.status) {
      case "blocked":
      case "error": {
        return {
          key: null,
          fetcher: () => {
            throw new Error("fetcher unexpectedly invoked");
          }
        };
      }
      case "ready": {
        const cacheKey = makeQueryCacheKey(
          query.id,
          paramsResult.resolvedParams
        );
        return {
          key: cacheKey,
          fetcher: () => {
            const clientCachedPromise = GLOBAL_CACHE.get(cacheKey);
            if (clientCachedPromise == null ? void 0 : clientCachedPromise.result) {
              if (clientCachedPromise.result.state === "resolved") {
                $queryState.resolvePromise(
                  cacheKey,
                  clientCachedPromise.result.value
                );
              } else {
                $queryState.rejectPromise(
                  cacheKey,
                  clientCachedPromise.result.error
                );
              }
              return clientCachedPromise.promise.finally(() => {
                GLOBAL_CACHE.delete(cacheKey);
              });
            } else {
              const promise = query.fn(...paramsResult.resolvedParams);
              $queryState.loadingPromise(cacheKey, promise);
              return promise.finally(() => {
                GLOBAL_CACHE.delete(cacheKey);
              });
            }
          }
        };
      }
    }
  }, [query, paramsResult]);
  const result = (0, import_query2.useMutablePlasmicQueryData)(key, fetcher, {
    // If revalidateIfStale is true, then if there's a cache entry with a key,
    // but no mounted hook with that key yet, and when the hook mounts with the key,
    // swr will revalidate. This may be reasonable behavior, but for us, this
    // happens all the time -- we prepopulate the cache with proxy-invoked fetch,
    // sometimes before swr had a chance to run the effect.  So we turn off
    // revalidateIfStale here, and just let the user manage invalidation.
    revalidateIfStale: false
  });
  if (!result.isLoading) {
    if (result.error) {
      $queryState.rejectPromise(key, result.error);
    } else if (key && result.data !== void 0) {
      $queryState.resolvePromise(key, result.data);
    }
  }
  return result;
}

// src/serverQueries/server.ts
var ROOT_COMPONENT_KEY_PATH = "root";
function appendKeyPath(currentKeyPath, currentInput) {
  return currentKeyPath ? `${currentKeyPath}/${currentInput}` : currentInput;
}
function executeQueryTree(rootNode, options, queriesByComponent) {
  const { $props, $ctx } = options;
  const initialContext = {
    $props,
    $ctx,
    // Placeholder; executeComponentNode replaces this with an initial
    // $state derived from the component's own stateSpecs.
    $state: {},
    $q: {},
    $scopedItemVars: {}
  };
  return executeComponentNode(rootNode, {
    context: initialContext,
    parentKeyPath: "",
    childIndex: 0,
    queriesByComponent
  });
}
function executeNode(node, params) {
  const p = { node, params, type: node.type };
  switch (p.type) {
    case "component":
      return executeComponentNode(p.node, p.params);
    case "codeComponent":
      return executeCodeComponentNode(p.node, p.params);
    case "dataProvider":
      return executeDataProviderNode(p.node, p.params);
    case "visibility":
      return executeVisibilityNode(p.node, p.params);
    case "repeated":
      return executeRepeatedNode(p.node, p.params);
    default:
      assertUnexpectedNodeType(p);
  }
}
function executeComponentNode(node, params) {
  const {
    context: parentContext,
    parentKeyPath,
    childIndex,
    queriesByComponent
  } = params;
  const isRoot = parentKeyPath === "";
  let componentKeyPath;
  let evaluatedProps;
  if (isRoot) {
    componentKeyPath = ROOT_COMPONENT_KEY_PATH;
    evaluatedProps = parentContext.$props;
  } else {
    componentKeyPath = appendKeyPath(parentKeyPath, `c${childIndex}`);
    evaluatedProps = {};
    for (const [propName, propFn] of Object.entries(node.propsContext)) {
      const result = safeExecResult(() => propFn(parentContext));
      if (!("data" in result)) {
        return [];
      }
      evaluatedProps[propName] = result.data;
    }
  }
  let componentQueries = queriesByComponent.get(componentKeyPath);
  if (!componentQueries) {
    componentQueries = {};
    queriesByComponent.set(componentKeyPath, componentQueries);
  }
  const $state = node.stateSpecs.length > 0 ? createInitial$State(
    parentContext.$ctx,
    evaluatedProps,
    componentQueries,
    node.stateSpecs
  ) : {};
  const componentContext = {
    $props: evaluatedProps,
    $ctx: parentContext.$ctx,
    $state,
    $q: componentQueries,
    $scopedItemVars: parentContext.$scopedItemVars
  };
  const discovered = [];
  for (const [queryName, query] of Object.entries(node.queries)) {
    if (componentQueries[queryName]) {
      continue;
    }
    const $query = new StatefulQueryResult();
    componentQueries[queryName] = $query;
    discovered.push({ $query, query, ctx: componentContext });
  }
  node.children.forEach((child, idx) => {
    discovered.push(
      ...executeNode(child, {
        context: componentContext,
        parentKeyPath: componentKeyPath,
        childIndex: idx,
        queriesByComponent
      })
    );
  });
  return discovered;
}
function executeCodeComponentNode(node, params) {
  if (node.serverRenderingConfig === false) {
    return [];
  }
  const { context, parentKeyPath, childIndex, queriesByComponent } = params;
  const nodeKeyPath = appendKeyPath(parentKeyPath, `cc${childIndex}`);
  const evaluatedProps = {};
  for (const [propName, propFn] of Object.entries(node.propsContext)) {
    const result = safeExecResult(() => propFn(context));
    if ("data" in result) {
      evaluatedProps[propName] = result.data;
    } else {
      return [];
    }
  }
  const childContext = {
    $props: evaluatedProps,
    $ctx: context.$ctx,
    $state: context.$state,
    $q: context.$q,
    $scopedItemVars: context.$scopedItemVars
  };
  return node.children.flatMap(
    (child, idx) => executeNode(child, {
      context: childContext,
      parentKeyPath: nodeKeyPath,
      childIndex: idx,
      queriesByComponent
    })
  );
}
function executeDataProviderNode(node, params) {
  const { context, parentKeyPath, childIndex, queriesByComponent } = params;
  const nodeKeyPath = appendKeyPath(parentKeyPath, `dp${childIndex}`);
  const dataResult = safeExecResult(() => node.data(context));
  if (!("data" in dataResult)) {
    return [];
  }
  const childContext = {
    $props: context.$props,
    $ctx: __spreadProps(__spreadValues({}, context.$ctx), {
      [node.name]: dataResult.data
    }),
    $state: context.$state,
    $q: context.$q,
    $scopedItemVars: context.$scopedItemVars
  };
  return node.children.flatMap(
    (child, idx) => executeNode(child, {
      context: childContext,
      parentKeyPath: nodeKeyPath,
      childIndex: idx,
      queriesByComponent
    })
  );
}
function executeVisibilityNode(node, params) {
  const { context, parentKeyPath, childIndex, queriesByComponent } = params;
  const nodeKeyPath = appendKeyPath(parentKeyPath, `v${childIndex}`);
  const visibilityResult = safeExecResult(() => node.visibilityExpr(context));
  if (!("data" in visibilityResult) || !visibilityResult.data) {
    return [];
  }
  return node.children.flatMap(
    (child, idx) => executeNode(child, {
      context,
      parentKeyPath: nodeKeyPath,
      childIndex: idx,
      queriesByComponent
    })
  );
}
function executeRepeatedNode(node, params) {
  const { context, parentKeyPath, childIndex, queriesByComponent } = params;
  const nodeKeyPath = appendKeyPath(parentKeyPath, `r${childIndex}`);
  const collectionResult = safeExecResult(() => node.collectionExpr(context));
  if (!("data" in collectionResult) || !Array.isArray(collectionResult.data)) {
    return [];
  }
  return collectionResult.data.flatMap((item, index) => {
    const itemContext = {
      $props: context.$props,
      $ctx: context.$ctx,
      $state: context.$state,
      $q: context.$q,
      $scopedItemVars: __spreadProps(__spreadValues({}, context.$scopedItemVars), {
        [node.itemName]: item,
        [node.indexName]: index
      })
    };
    const itemKeyPath = appendKeyPath(nodeKeyPath, `i${index}`);
    return node.children.flatMap(
      (child, idx) => executeNode(child, {
        context: itemContext,
        parentKeyPath: itemKeyPath,
        childIndex: idx,
        queriesByComponent
      })
    );
  });
}
function executePlasmicQueries(rootNode, options) {
  return __async(this, null, function* () {
    const queriesByComponent = /* @__PURE__ */ new Map();
    const discoveredQueries = [];
    while (true) {
      const newQueries = executeQueryTree(rootNode, options, queriesByComponent);
      if (newQueries.length === 0) {
        break;
      }
      discoveredQueries.push(...newQueries);
      yield Promise.all(
        newQueries.map(
          (d) => executePlasmicQuery(d.$query, d.query, d.ctx).catch(() => {
          })
        )
      );
    }
    const cache = {};
    for (const d of discoveredQueries) {
      if ("data" in d.$query.current) {
        cache[d.$query.current.key] = d.$query.current.data;
      }
    }
    const queries = {};
    const rootQueries = queriesByComponent.get(ROOT_COMPONENT_KEY_PATH);
    if (rootQueries) {
      for (const [name, $query] of Object.entries(rootQueries)) {
        if ("data" in $query.current) {
          queries[name] = {
            key: $query.key,
            data: $query.current.data,
            isLoading: false
          };
        }
      }
    }
    for (const d of discoveredQueries) {
      if ("error" in d.$query.current) {
        throw d.$query.current.error;
      }
    }
    return { cache, queries };
  });
}
function executePlasmicQuery($query, query, ctx) {
  return __async(this, null, function* () {
    if ($query.current.state === "loading" || $query.current.state === "done") {
      return $query.getDoneResult();
    }
    do {
      const paramsResult = resolveParams(query.id, () => query.args(ctx));
      switch (paramsResult.status) {
        case "blocked": {
          try {
            yield paramsResult.promise;
          } catch (e) {
          }
          continue;
        }
        case "ready": {
          $query.loadingPromise(
            paramsResult.cacheKey,
            Promise.resolve(query.fn(...paramsResult.resolvedParams))
          );
          return $query.getDoneResult();
        }
        case "error": {
          $query.rejectPromise(null, paramsResult.error);
          throw paramsResult.error;
        }
      }
    } while (true);
  });
}
function executeServerQuery(_query) {
  return __async(this, null, function* () {
    return { data: void 0, isLoading: false };
  });
}

// src/index.tsx
var import_query5 = require("@plasmicapp/query");

// src/components/Fetcher.tsx
var import_react3 = __toESM(require("react"));

// src/hooks/usePlasmicDataOp.tsx
var import_data_sources_context = require("@plasmicapp/data-sources-context");
var import_query4 = require("@plasmicapp/query");
var React4 = __toESM(require("react"));

// src/executor.tsx
var import_isomorphic_unfetch = __toESM(require("@plasmicapp/isomorphic-unfetch"));
var import_query3 = require("@plasmicapp/query");
var import_fast_stringify = __toESM(require("fast-stringify"));

// src/placeholders.ts
var PLASMIC_UNDEFINED = "__PLASMIC_UNDEFINED";
function addPlaceholders(val) {
  return val === void 0 ? PLASMIC_UNDEFINED : val;
}
function addPlaceholdersToUserArgs(userArgs) {
  if (!userArgs) {
    return userArgs;
  }
  Object.entries(userArgs).forEach(([key, val]) => {
    userArgs[key] = Array.isArray(val) ? val.map((v) => addPlaceholders(v)) : addPlaceholders(val);
  });
  return userArgs;
}

// src/executor.tsx
var DEFAULT_HOST = "https://data.plasmic.app";
var UNAUTHORIZED_MESSAGE = "You do not have permission to perform this operation. Login to get access or contact the app owner to get access.";
function executePlasmicDataOp(op, opts) {
  return __async(this, null, function* () {
    const func = getConfig(
      "__PLASMIC_EXECUTE_DATA_OP",
      _executePlasmicDataOp
    );
    op.userArgs = addPlaceholdersToUserArgs(op.userArgs);
    const res = yield (0, import_query3.wrapLoadingFetcher)(func)(op, opts);
    return res;
  });
}
function _executePlasmicDataOp(op, opts) {
  return __async(this, null, function* () {
    var _a;
    if (op.roleId) {
      if (!(opts == null ? void 0 : opts.user) || !opts.user.roleIds.includes(op.roleId)) {
        console.error(UNAUTHORIZED_MESSAGE);
        throw new Error(UNAUTHORIZED_MESSAGE);
      }
    }
    const host = getConfig("__PLASMIC_DATA_HOST", DEFAULT_HOST);
    const url = `${host}/api/v1/server-data/sources/${op.sourceId}/execute`;
    const resp = yield (0, import_isomorphic_unfetch.default)(url, {
      method: "POST",
      headers: __spreadValues({
        "Content-Type": "application/json"
      }, (opts == null ? void 0 : opts.userAuthToken) && {
        "x-plasmic-data-user-auth-token": opts.userAuthToken
      }),
      body: (0, import_fast_stringify.default)({
        opId: op.opId,
        userArgs: (_a = op.userArgs) != null ? _a : {},
        paginate: opts == null ? void 0 : opts.paginate
      })
    });
    if (resp.status !== 200) {
      const text = yield resp.text();
      throw new Error(text);
    }
    return yield resp.json();
  });
}
function getConfig(key, defaultValue) {
  var _a;
  if (typeof globalThis === "undefined") {
    return defaultValue;
  } else {
    return (_a = globalThis[key]) != null ? _a : defaultValue;
  }
}

// src/hooks/usePlasmicDataOp.tsx
function makeCacheKey(dataOp, opts) {
  const queryDependencies = JSON.stringify({
    sourceId: dataOp.sourceId,
    opId: dataOp.opId,
    args: dataOp.userArgs,
    userAuthToken: opts == null ? void 0 : opts.userAuthToken,
    paginate: opts == null ? void 0 : opts.paginate
  });
  return dataOp.cacheKey ? `${dataOp.cacheKey}${queryDependencies}` : queryDependencies;
}
function usePlasmicInvalidate() {
  const { cache, fallback, mutate } = (0, import_query4.usePlasmicDataConfig)();
  return (invalidatedKeys) => __async(null, null, function* () {
    const getKeysToInvalidate = () => {
      var _a, _b;
      if (!invalidatedKeys) {
        return [];
      }
      const allKeys = Array.from(
        /* @__PURE__ */ new Set([
          ...Array.from(cache.keys()),
          ...fallback ? Object.keys(fallback) : [],
          // If this is running within the Studio, we also take the
          // opportunity to invalidate the Studio cache. The keys that
          // Studio may have can be a superset of `cache` here, because
          // `cache` is updated as swr hooks are mounted and unmounted,
          // but Studio's data cache keys don't get removed when hooks
          // are unmounted. This makes it possible for Studio to hold
          // onto a stale cache entry that doesn't get invalidated.
          // For example, Studio may render page1, with key X, then goes
          // to page 2, which performs a mutate. At this point, Studio
          // has a cache entry for key X, but `cache` does not, because
          // page2 does not use that query. But page 2 may perform a
          // mutation that invalidates X. So we need to invalidate not
          // only keys in `cache`, but also keys that Studio is still
          // holding onto.
          ...(_b = (_a = globalThis.__PLASMIC_GET_ALL_CACHE_KEYS) == null ? void 0 : _a.call(globalThis)) != null ? _b : []
        ])
      ).filter((key) => typeof key === "string");
      if (invalidatedKeys.includes("plasmic_refresh_all")) {
        return allKeys;
      }
      return allKeys.filter(
        (key) => invalidatedKeys.some((k) => key.includes(`.$.${k}.$.`))
      );
    };
    const keys = getKeysToInvalidate();
    if (keys.length === 0) {
      return;
    }
    const invalidateKey = (key) => __async(null, null, function* () {
      const studioInvalidate = globalThis.__PLASMIC_MUTATE_DATA_OP;
      if (studioInvalidate) {
        yield studioInvalidate(key);
      }
      return mutate(key);
    });
    return yield Promise.all(keys.map((key) => invalidateKey(key)));
  });
}
function resolveDataOp(dataOp) {
  if (typeof dataOp === "function") {
    try {
      return dataOp();
    } catch (err) {
      if (isPlasmicUndefinedDataErrorPromise(err)) {
        return err;
      }
      return null;
    }
  } else {
    return dataOp;
  }
}
function usePlasmicDataOp(dataOp, opts) {
  const resolvedDataOp = resolveDataOp(dataOp);
  const ctx = (0, import_data_sources_context.usePlasmicDataSourceContext)();
  const key = !resolvedDataOp || isPlasmicUndefinedDataErrorPromise(resolvedDataOp) ? null : makeCacheKey(resolvedDataOp, {
    paginate: opts == null ? void 0 : opts.paginate,
    userAuthToken: ctx == null ? void 0 : ctx.userAuthToken
  });
  const fetcher = (op) => {
    return executePlasmicDataOp(op, {
      userAuthToken: (ctx == null ? void 0 : ctx.userAuthToken) || void 0,
      user: ctx == null ? void 0 : ctx.user,
      paginate: opts == null ? void 0 : opts.paginate
    });
  };
  const resultMapper = (result) => {
    var _a;
    return __spreadValues(__spreadValues({}, (_a = result.data) != null ? _a : {}), pick(result, "error", "isLoading"));
  };
  return usePlasmicFetch(
    key,
    resolvedDataOp,
    fetcher,
    resultMapper,
    ["data", "schema", "error"],
    {
      noUndefinedDataProxy: opts == null ? void 0 : opts.noUndefinedDataProxy
    }
  );
}
function usePlasmicDataMutationOp(dataOp) {
  const ctx = (0, import_data_sources_context.usePlasmicDataSourceContext)();
  const userToken = ctx == null ? void 0 : ctx.userAuthToken;
  const getRealDataOp = React4.useCallback(() => __async(null, null, function* () {
    const tryGetRealDataOp = () => __async(null, null, function* () {
      const resolved = resolveDataOp(dataOp);
      if (!resolved) {
        return null;
      } else if (isPlasmicUndefinedDataErrorPromise(resolved)) {
        yield resolved;
        return tryGetRealDataOp();
      } else {
        return resolved;
      }
    });
    return yield tryGetRealDataOp();
  }), [dataOp]);
  return React4.useCallback(() => __async(null, null, function* () {
    var _a;
    const { sourceId, opId, userArgs } = (_a = yield getRealDataOp()) != null ? _a : {};
    if (!sourceId || !opId) {
      return void 0;
    }
    return executePlasmicDataOp(
      { sourceId, opId, userArgs },
      {
        userAuthToken: userToken || void 0,
        user: ctx == null ? void 0 : ctx.user
      }
    );
  }), [getRealDataOp, userToken]);
}

// src/components/Fetcher.tsx
function Fetcher(props) {
  var _a;
  const { dataOp, children, name, pageIndex, pageSize } = props;
  const data = usePlasmicDataOp(dataOp, __spreadValues({}, !!pageIndex && !!pageSize && {
    paginate: { pageIndex, pageSize }
  }));
  const $queries = import_react3.default.useMemo(
    () => __spreadProps(__spreadValues({}, props.queries), { [name != null ? name : "data"]: data }),
    [props.queries, name, data]
  );
  return (_a = children == null ? void 0 : children($queries)) != null ? _a : null;
}
var FetcherMeta = {
  name: "plasmic-data-source-fetcher",
  displayName: "Data Fetcher",
  props: {
    dataOp: {
      type: "dataSourceOp",
      displayName: "Data"
    },
    name: {
      type: "string",
      displayName: "Variable name"
    },
    children: {
      type: "slot",
      renderPropParams: ["$queries"]
    },
    pageSize: {
      type: "number",
      advanced: true,
      displayName: "Page size",
      description: "Only fetch in batches of this size; for pagination"
    },
    pageIndex: {
      type: "number",
      advanced: true,
      displayName: "Page index",
      description: "0-based index of the paginated page to fetch"
    }
  },
  importPath: "@plasmicapp/react-web/lib/data-sources",
  importName: "Fetcher",
  alwaysAutoName: true,
  styleSections: false
};

// src/helpers.ts
var import_react4 = require("react");
function normalizeData(rawData) {
  var _a;
  if (!rawData) {
    return void 0;
  }
  const dataArray = tryGetDataArray(rawData);
  if (!dataArray) {
    return void 0;
  }
  const schema = (_a = rawData.schema) != null ? _a : tryGetSchema(dataArray);
  if (!schema) {
    return void 0;
  }
  return { data: dataArray, schema };
}
function useNormalizedData(rawData) {
  return (0, import_react4.useMemo)(() => normalizeData(rawData), [rawData]);
}
function tryGetDataArray(rawData) {
  if (rawData == null || typeof rawData !== "object") {
    return void 0;
  }
  if (Array.isArray(rawData)) {
    if (isArrayOfObjects(rawData)) {
      return rawData;
    } else {
      return void 0;
    }
  }
  if (rawData == null) {
    return void 0;
  }
  if ("data" in rawData && typeof rawData.data === "object") {
    if (Array.isArray(rawData.data) && isArrayOfObjects(rawData.data)) {
      return rawData.data;
    } else if (rawData.data != null) {
      return [rawData.data];
    } else {
      return void 0;
    }
  }
  if ("isLoading" in rawData || "error" in rawData) {
    return void 0;
  }
  return [rawData];
}
function isArrayOfObjects(arr) {
  return arr.every((x) => typeof x === "object" && !Array.isArray(x));
}
function tryGetSchema(data) {
  const fieldMap = {};
  data.forEach((entry) => {
    if (entry && typeof entry === "object") {
      Array.from(Object.entries(entry)).forEach(([k, v]) => {
        const inferredType = typeof v === "string" ? "string" : typeof v === "boolean" ? "boolean" : typeof v === "number" ? "number" : "unknown";
        if (fieldMap[k] && fieldMap[k] !== inferredType) {
          fieldMap[k] = "unknown";
        } else {
          fieldMap[k] = inferredType;
        }
      });
    }
  });
  return {
    id: "inferred",
    fields: Object.entries(fieldMap).map(([f, t]) => ({
      id: f,
      type: t,
      readOnly: false
    }))
  };
}
var mkShortId = () => `${Math.random()}`;
function deriveFieldConfigs(specifiedFieldsPartial, schema, makeDefaultConfig) {
  var _a;
  const schemaFields = (_a = schema == null ? void 0 : schema.fields) != null ? _a : [];
  const fieldById = mkIdMap(schemaFields);
  const specifiedFieldIds = new Set(
    withoutNils(specifiedFieldsPartial.map((f) => f.fieldId))
  );
  const keptSpecifiedFields = specifiedFieldsPartial.flatMap((f) => {
    if (!f.fieldId) {
      return [
        __spreadValues(__spreadValues({ key: mkShortId() }, makeDefaultConfig(void 0)), f)
      ];
    }
    const field = fieldById.get(f.fieldId);
    if (!field) {
      return [];
    }
    return [
      __spreadValues(__spreadValues({
        key: mkShortId()
      }, makeDefaultConfig(field)), f)
    ];
  });
  const newVirtualFields = schemaFields.filter((f) => !specifiedFieldIds.has(f.id)).map(
    (f) => __spreadValues({
      key: mkShortId()
    }, makeDefaultConfig(f))
  );
  const mergedFields = [...keptSpecifiedFields, ...newVirtualFields];
  const minimalFullLengthFields = [
    ...specifiedFieldsPartial,
    ...newVirtualFields.map((f) => ({ key: f.key, fieldId: f.fieldId }))
  ];
  return { mergedFields, minimalFullLengthFields };
}
