// @plasmicapp/react-web/skinny/dist/ssr-6a83562a.js
import { a as __assign, o as omit, d as __spreadArray, e as __read, p as pick } from './common-a21bdc95.js';
import { PlasmicDataSourceContextProvider } from '@plasmicapp/data-sources-context';
import { usePlasmicLinkMaybe, PlasmicTranslatorContext as PlasmicTranslatorContext$1, usePlasmicTranslator as usePlasmicTranslator$1, DataProvider, PlasmicLinkProvider } from '@plasmicapp/host';
import { SSRProvider, useIsSSR as useIsSSR$1 } from '@react-aria/ssr';
import * as React from 'react';
import React__default from 'react';
import { PlasmicHeadContext } from './render/PlasmicHead/index.js';

// These props are used internally by PlasmicLink to determine how to render the link
var INTERNAL_PROPS = ["component", "platform", "legacyBehavior"];
var PlasmicLink = React__default.forwardRef(function PlasmicLink(props, ref) {
    var _a;
    // The usePlasmicLinkMaybe function may be undefined, if host is not up-to-date
    var Link = (_a = usePlasmicLinkMaybe === null || usePlasmicLinkMaybe === void 0 ? void 0 : usePlasmicLinkMaybe()) !== null && _a !== void 0 ? _a : PlasmicLinkInternal;
    if (Link === PlasmicLink || Link === PlasmicLinkInternal) {
        // Just in case, break the cycle
        return React__default.createElement(PlasmicLinkInternal, __assign({}, props, { ref: ref }));
    }
    else {
        // Don't pass internal props (component/platform/legacyBehavior) to non-PlasmicLinkInternal
        // eslint-disable-next-line react/forbid-elements
        return React__default.createElement(Link, __assign({}, omit.apply(void 0, __spreadArray([props], __read(INTERNAL_PROPS), false)), { ref: ref }));
    }
});
var PlasmicLinkInternal = React__default.forwardRef(function PlasmicLinkInternal(props, ref) {
    var _a;
    // props.href is required for nextjs; if no props.href,
    // then we just render the default anchor element
    if (props.platform === "nextjs" && props.href) {
        var nextjsProps = [
            "href",
            "replace",
            "scroll",
            "shallow",
            "passHref",
            "prefetch",
            "locale",
        ];
        // If this is a fragment identifier link, then we set
        // scroll={false} so that smooth scrolling works
        var isFragment = (_a = props.href) === null || _a === void 0 ? void 0 : _a.startsWith("#");
        // Default to legacy behavior (nested <a> tag) unless legacyBehavior
        // is explicitly set to false. Older codegen may not set this prop at
        // all, so we treat undefined as legacy for backwards compatibility.
        // https://github.com/plasmicapp/plasmic-internal/pull/2203#issuecomment-3877931788
        if (props.legacyBehavior !== false) {
            return React__default.createElement(props.component, __assign({ scroll: !isFragment, legacyBehavior: true }, pick.apply(void 0, __spreadArray([props], __read(nextjsProps), false))), React__default.createElement("a", __assign({}, omit.apply(void 0, __spreadArray(__spreadArray([props], __read(INTERNAL_PROPS), false), __read(nextjsProps), false)), { ref: ref })));
        }
        // pass props directly to Link (no nested <a>)
        return React__default.createElement(props.component, __assign(__assign({ scroll: !isFragment }, omit.apply(void 0, __spreadArray([props], __read(INTERNAL_PROPS), false))), { ref: ref }));
    }
    if (props.platform === "gatsby" && isInternalHref(props.href)) {
        return React__default.createElement(props.component, __assign(__assign({}, omit.apply(void 0, __spreadArray(__spreadArray([props], __read(INTERNAL_PROPS), false), ["href"], false))), { to: props.href, ref: ref }));
    }
    if (props.platform === "tanstack" && isInternalHref(props.href)) {
        return React__default.createElement(props.component, __assign(__assign({}, omit.apply(void 0, __spreadArray(__spreadArray([props], __read(INTERNAL_PROPS), false), ["href"], false))), { to: props.href, ref: ref }));
    }
    return React__default.createElement("a", __assign({}, omit.apply(void 0, __spreadArray([props], __read(INTERNAL_PROPS), false)), { ref: ref }));
});
function isInternalHref(href) {
    return /^\/(?!\/)/.test(href);
}

// Make the refactor to host backwards compatible for loader
var PlasmicTranslatorContext = PlasmicTranslatorContext$1 !== null && PlasmicTranslatorContext$1 !== void 0 ? PlasmicTranslatorContext$1 : React__default.createContext(undefined);
var usePlasmicTranslator = usePlasmicTranslator$1 !== null && usePlasmicTranslator$1 !== void 0 ? usePlasmicTranslator$1 : (function () {
    var _t = React__default.useContext(PlasmicTranslatorContext);
    var translator = _t
        ? typeof _t === "function"
            ? _t
            : _t.translator
        : undefined;
    return translator;
});
function genTranslatableString(elt, opts) {
    var components = {};
    var componentsCount = 0;
    var getText = function (node) {
        var _a;
        if (!node) {
            return "";
        }
        if (typeof node === "number" ||
            typeof node === "boolean" ||
            typeof node === "string") {
            return node.toString();
        }
        if (typeof node !== "object") {
            return "";
        }
        if (Array.isArray(node) || isIterable(node)) {
            return Array.from(node)
                .map(function (child) { return getText(child); })
                .filter(function (child) { return !!child; })
                .join("");
        }
        var nodeChildren = (hasKey(node, "props") &&
            hasKey(node.props, "children") &&
            node.props.children) ||
            (hasKey(node, "children") && node.children) ||
            [];
        var contents = "".concat(React__default.Children.toArray(nodeChildren)
            .map(function (child) { return getText(child); })
            .filter(function (child) { return !!child; })
            .join(""));
        if (React__default.isValidElement(node) && node.type === React__default.Fragment) {
            return contents;
        }
        var prefix = (_a = opts === null || opts === void 0 ? void 0 : opts.tagPrefix) !== null && _a !== void 0 ? _a : "";
        var componentId = "".concat(prefix).concat(componentsCount + 1);
        componentsCount++;
        components[componentId] = React__default.isValidElement(node)
            ? React__default.cloneElement(node, {
                key: componentId,
                children: undefined,
            })
            : node;
        return "<".concat(componentId, ">").concat(contents, "</").concat(componentId, ">");
    };
    var str = getText(elt);
    return {
        str: str,
        components: components,
        componentsCount: componentsCount,
    };
}
function Trans(_a) {
    var transKey = _a.transKey, children = _a.children;
    var _t = React__default.useContext(PlasmicTranslatorContext);
    var translator = _t
        ? typeof _t === "function"
            ? _t
            : _t.translator
        : undefined;
    if (!translator) {
        warnNoTranslationFunctionAtMostOnce();
        return children;
    }
    var _b = genTranslatableString(children, {
        tagPrefix: typeof _t === "object" ? _t.tagPrefix : undefined,
    }), str = _b.str, components = _b.components, componentsCount = _b.componentsCount;
    return translator(transKey !== null && transKey !== void 0 ? transKey : str, componentsCount > 0 ? { components: components } : undefined);
}
var hasWarned = false;
function warnNoTranslationFunctionAtMostOnce() {
    if (!hasWarned) {
        console.warn("Using Plasmic Translation but no translation function has been provided");
        hasWarned = true;
    }
}
function hasKey(v, key) {
    return typeof v === "object" && v !== null && key in v;
}
function isIterable(val) {
    return val != null && typeof val[Symbol.iterator] === "function";
}

var PlasmicRootContext = React.createContext(undefined);
/**
 * PlasmicRootProvider sets up the React context that Plasmic-generated components
 * rely on including data sources, i18n, Head, and Link.
 *
 * In Next.js app router, props passed from a Server to a Client Components must be serializable
 * but several PlasmicRootProvider props are not (e.g. `loader`, `Link` from `next/link`).
 * We recommend defining a Client Component wrapper (`ClientPlasmicRootProvider` in
 * `plasmic-init-client.tsx`) that imports non-serializable values and passes them to
 * PlasmicRootProvider (and only accepts serializable props from its caller).
 *
 * Loader example:
 *
 * ```tsx
 * // plasmic-init-client.tsx
 * "use client";
 * import { PlasmicRootProvider } from "@plasmicapp/loader-nextjs";
 * import { PLASMIC } from "@/plasmic-init";
 * export function ClientPlasmicRootProvider(
 *   props: Omit<React.ComponentProps<typeof PlasmicRootProvider>, "loader">
 * ) {
 *   return <PlasmicRootProvider loader={PLASMIC} {...props} />;
 * }
 * ```
 *
 * Codegen example:
 *
 * ```tsx
 * // plasmic-init-client.tsx
 * "use client";
 * import { PlasmicRootProvider } from "@plasmicapp/react-web";
 * import Link from "next/link";
 * export function ClientPlasmicRootProvider(
 *   props: Omit<React.ComponentProps<typeof PlasmicRootProvider>, "Link">
 * ) {
 *   return <PlasmicRootProvider Link={Link} {...props} />;
 * }
 * ```
 *
 * A Server Component can then render `ClientPlasmicRootProvider` and pass
 * serializable props such as prefetched data and children:
 *
 * ```tsx
 * import { PLASMIC } from "@/plasmic-init";
 * import { ClientPlasmicRootProvider } from "@/plasmic-init-client";
 * export default async function MyPage() {
 *   const prefetchedData = await PLASMIC.fetchComponentData("YourPage");
 *   return (
 *     <ClientPlasmicRootProvider prefetchedData={prefetchedData}>
 *       {yourContent()}
 *     </ClientPlasmicRootProvider>
 *   );
 * }
 * ```
 *
 * See https://nextjs.org/docs/app/getting-started/server-and-client-components#passing-data-from-server-to-client-components
 * for more on the Server/Client Component boundary.
 */
function PlasmicRootProvider(props) {
    var _a;
    var platform = props.platform, children = props.children, userAuthToken = props.userAuthToken, isUserLoading = props.isUserLoading, authRedirectUri = props.authRedirectUri, user = props.user, disableLoadingBoundary = props.disableLoadingBoundary, suspenseFallback = props.suspenseFallback;
    var context = React.useMemo(function () { return ({
        platform: platform,
    }); }, [platform]);
    var dataSourceContextValue = React.useMemo(function () { return ({
        userAuthToken: userAuthToken,
        user: user,
        isUserLoading: isUserLoading,
        authRedirectUri: authRedirectUri,
    }); }, [userAuthToken, isUserLoading, user, authRedirectUri]);
    var reactMajorVersion = +React.version.split(".")[0];
    return (React.createElement(MaybeWrap, { cond: !disableLoadingBoundary && reactMajorVersion >= 18, wrapper: function (children_) { return (React.createElement(DataProvider, { name: "plasmicInternalEnableLoadingBoundary", hidden: true, data: true },
            React.createElement(React.Suspense, { fallback: suspenseFallback !== null && suspenseFallback !== void 0 ? suspenseFallback : "Loading..." }, children_))); } },
        React.createElement(PlasmicRootContext.Provider, { value: context },
            React.createElement(MaybeWrap, { cond: reactMajorVersion < 18, wrapper: function (children_) { return React.createElement(SSRProvider, null, children_); } },
                React.createElement(PlasmicDataSourceContextProvider, { value: dataSourceContextValue },
                    React.createElement(PlasmicTranslatorContext.Provider, { value: props.i18n },
                        React.createElement(PlasmicHeadContext.Provider, { value: props.Head },
                            React.createElement(SafePlasmicLinkProvider, { Link: (_a = props.Link) !== null && _a !== void 0 ? _a : PlasmicLinkInternal }, children))))))));
}
/**
 * A PlasmicLinkProvider that anticipates PlasmicLinkProvider may not exist yet from
 * @plasmicapp/host if the user is using an older version
 */
function SafePlasmicLinkProvider(props) {
    if (PlasmicLinkProvider) {
        return React.createElement(PlasmicLinkProvider, __assign({}, props));
    }
    else {
        return React.createElement(React.Fragment, null, props.children);
    }
}
var useIsSSR = useIsSSR$1;
function useHasPlasmicRoot() {
    return !!React.useContext(PlasmicRootContext);
}
var hasWarnedSSR = false;
/**
 * Warns the user if PlasmicRootProvider is not used
 */
function useEnsureSSRProvider() {
    var hasRoot = useHasPlasmicRoot();
    if (hasRoot || hasWarnedSSR || process.env.NODE_ENV !== "development") {
        return;
    }
    hasWarnedSSR = true;
    console.warn("Plasmic: To ensure your components work correctly with server-side rendering, please use PlasmicRootProvider at the root of your application.  See https://docs.plasmic.app/learn/ssr");
}
function MaybeWrap(props) {
    return props.cond ? props.wrapper(props.children) : props.children;
}

export { PlasmicLink as P, Trans as T, PlasmicRootProvider as a, usePlasmicTranslator as b, useEnsureSSRProvider as c, genTranslatableString as g, useIsSSR as u };
